x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/job/DataFrameJobTask.java

```diff
@@ -204,22 +210,59 @@ synchronized void shutdown() {
     @Override
     public synchronized void onCancelled() {
         logger.info(
-                "Received cancellation request for data frame job [" + job.getConfig().getId() + "], state: [" + indexer.getState() + "]");
+                "Received cancellation request for data frame job [" + job.getId() + "], state: [" + indexer.getState() + "]");
         if (indexer.abort()) {
             // there is no background job running, we can shutdown safely
             shutdown();
         }
     }
 
     protected class ClientDataFrameIndexer extends DataFrameIndexer {
+        private static final int LOAD_JOB_TIMEOUT_IN_SECONDS = 30;
         private final Client client;
+        private final DataFrameJobConfigManager jobConfigManager;
+        private final String jobId;
+
+        private DataFrameJobConfig jobConfig = null;
 
-        public ClientDataFrameIndexer(DataFrameJob job, AtomicReference<IndexerState> initialState,
+        public ClientDataFrameIndexer(String jobId, DataFrameJobConfigManager jobConfigManager, AtomicReference<IndexerState> initialState,
                 Map<String, Object> initialPosition, Client client) {
-            super(threadPool.executor(ThreadPool.Names.GENERIC), job, initialState, initialPosition);
+            super(threadPool.executor(ThreadPool.Names.GENERIC), initialState, initialPosition);
+            this.jobId = jobId;
+            this.jobConfigManager = jobConfigManager;
             this.client = client;
         }
 
+        @Override
+        protected DataFrameJobConfig getConfig() {
+            return jobConfig;
+        }
+
+        @Override
+        protected String getJobId() {
+            return jobId;
+        }
+
+        @Override
+        public synchronized boolean maybeTriggerAsyncJob(long now) {
+            if (jobConfig == null) {
+                CountDownLatch latch = new CountDownLatch(1);
```

Comment:
I am not sure how this latch is ever counted down. Maybe you should use a `LatchedActionListener`?